### DISCUSSION TITLE
Testing facilities Discussions

### DISCUSSION THREAD

**Discussion by mcuee** [avrdudes/avrdude#1021] [General]
It will be good to create a list of testing facilities for the volunteer testers: OS, Programmers and AVR chips.

OS: Windows (**not so sure who have ARM64 Windows machine**), Linux (x86_64, ARM64), macOS (Intel or ARM64), FreeBSD, etc

Programmers: Microchip programmers supported by Microchip for AVR chips (PICKit 2 is not one of them); non-Microchip programmers

AVR chips or development boards: Microchip official boards will usually have the programmer built-in. Arduino boards will usually have the bootloaders.

**Comment by mcuee** [comment]
For me I have access to Intel x86_64 Windows 10/11; Mac Mini M1 (ARM64) and Linux (Intel x86_64 and ARM64 with Raspberry Pi 400). I can also test FreeBSD, OpenBSD and NetBSD using virtual machines inside Windows or Proxmox V7.4 or v8 (low power Intel CPU based mini PC).

The following tools are as on 23-Oct-2023 --> 10-June-2025

As for programmers, I have the following.
1) avrftdi -- FT2232H based JTAGkey2, **courtesy of Amontec many years ago**
2) ft245r -- FT232R serial to TTL converter based programmer (arduino-ft232r)
3) PICKit 2 -- two units, **one unit coutesy of Microchip many years ago.**
4) usbasp -- Atmega8A or ATmega88 based; a few of the ATmega8A version has the UART pins on the ISP connectors. And the ATmega8A version can be used for testing of usbasploader as well.
5) usbtinyisp -- ATtiny2313 or ATtiny44 based
6) AVR Dragon (JTAGmkii protocol)-- official Microchip AVR Dragon, **courtesy of Atmel many years ago**
7) jtagmki: AVR JTAG ICE clone
8) stk500v2: AVRISP mkII clone (which works under Atmel Studio 7)
9) stk500v2: STK500v2 clone with CH340, just the programming portion
10) stk500v1/stk500v2: STK500 clone with ATmega8535 and supposedly capable of high voltage programming. Compatible with stk500v1/v2 when using V1 and V2 FW respectively.
11) stk500v2: STK500v2 clone with STC8051 MCU, supposedly capable of high voltage programming (not tested yet the high voltage side)
12) Microchip AVR128DB48 CURIOSITY NANO (with built-in pkobn_updi programmer)
13) Microchip ATMEGA328PB XPLAINED MINI (with built-in xplainedmini programmer)
14) Arduino Nano Every official board with ATmega4809 (with built-in jtag2updi programmer)
15) Arduino Nano ATmega4808 clone x 2pcs (with built-in jtag2updi programmer)
16) Raspberry Pi 3B+ and Rapberry Pi 400 with linuxspi/linuxgpio using an I/O extension board. 
17) Microchip SNAP
18) SI-Prog (serbb) -- it does not seem to work
19) Serial AVR910 programmer (ATtiny2313 based) -- this does not seem to work 
20) USB AVR910 (ATmega8535 based) -- this works fine
21) Microchip ATtiny817 Xplained Mini (with built-in xplainedmini_updi programmer)
22) serialupdi using an FT232RL USB to TTL converter + 1k ohm resistor
23) Microchip AVR64DD32 Curiosity Nano (with built-in pkobn_updi prorammer) (one dead but I got a new one)
24) Microchip ATtiny1627 Curiosity Nano (with built-in pkobn_updi prorammer)
25) Microchip PICKit 4 **(on loan from Microchip Singapore)**
26) Microchip ATtiny817 Xplained Pro **(on loan from Microchip Singapore)**
27) Microchip ATtiny416 Xplained Nano 
28) Microchip ATtiny104 Xplained Nano 
29) Microchip AVR64EA48 Curiosity Nano (with built-in pkobn_updi prorammer)
30) Nice serialupdi adapter with pogo pins, **courtesy of Spence Konde (aka Dr. Azzy)**
31) AVR JTAGICE mkII clone (programmer only, no debugging function) -- June 2025
32) STK600 programmer only clone -- June 2025

No plans to get as of now:
Microchip JTAG ICE 3 (obsolete), Atmel-ICE, Power Debugger

I have the following boards as well.
1) Arduino Leonardo official board with ATmega32U4 (with built-in avr109 bootloader)
2) Quite a number of Arduino Uno clones (ATmega328P) using ATmega16U2 or CH340G;  and Arduino Nano clones using CH340G (with built-in optiboot bootloader; also working with AVR109 bootloader)
3) Two Arduino Mega25600 clone (with built-in stk500v2 bootloader, or optiboot bootloader, also working with AVR109 bootloader using xboot)
4) One ATmega32A based board running Mighty Core (with built-in optiboot bootloader) (dead now but I have other ATmega32A board now)
5) Quite a few ATtiny88 based boards (with micronucleus bootloader)
6) ATmega128A core board with JTAG/ISP interface (not running any bootloader yet)
7) ATtiny85 boards and one ATtiny45 board (Arduino Gemma for ATtiny85 or micronucleus bootloader for both ATtiny45/85)
8) ATtiny13A boards (no bootloader -- but urclock bootloader works fine)
9) ATtiny167 boards (micronucleus bootloader)
10) ATxmega32D4 break-out board with PDI interface
11) ATmega8A, ATmega16A and ATmega32A breakout board (working with Optiboot and AVR109 using an external USB-TTL converter)
12) ATmega32A board with JTAG/ISP interface (working with optiboot and AVR109 bootloader)
13) Teensy 2.0 and 2.0++
14) ATxmeag32A4 breakout board with PDI interface 
15) LGT8F328P based Nano style boards
16) AVR64DD32 board, **courtesy of Spence Konde (aka Dr. Azzy)**
17) AVR32DD14 breakout board, **courtesy of Spence Konde (aka Dr. Azzy)**
18) ATTiny3227 breakout board, **courtesy of Spence Konde (aka Dr. Azzy**)
19) Arduino Uno Clone using ATmega328PB and CH340

**Reply by MCUdude** [reply]
These are the dedicated programmers/debuggers I own or have access to:
* AVRISPmkII
* AVR Dragon
* Atmel ICE
* Bus Pirate V3
* Curiosity Nano (UPDI)
* FT232H (arduino-ft232h)
* MPLAB SNAP
* JTAG ICE 3
* JTAG2UPDI
* PICkit2
* PICkit4
* Power Debugger
* SerialUPDI
* STK500 board (running STK500v2 firmware)
* STK600
* USBasp
* USBtinyISP
* Xplained Mini ISP
* Xplained Mini TPI
* Xplained Mini UPDI

These are the development boards worth mentioning
* Custom board for 32-pin classic ATmega8/88/168/328 on daughterboards, with onboard USB to serial adapter for use with Optiboot. 
* Custom board for 44-pin classic ATmega8535/16/32/164/324(PB)/644/1284 on daughterboards, with onboard USB to serial adapter for use with Optiboot.
* Custom board for use with DIP8, DIP14, and DIP20 classic ATtinys
* Custom boards for 32-pin megaAVR/AVRDx with onboard USB to serial adapter for use with Optiboot. 
* Custom boards for 48-pin megaAVR/AVRDx with onboard USB to serial adapter for use with Optiboot. 
* [MightyCore development board](https://www.tindie.com/products/mcudude/dip-40-arduino-compatible-development-board/?pt=ac_prod_search)
* Arduino Uno, Leonardo, Mega2560, Uno Wifi Rev2
* Xplained Mini Attiny817 and ATmega328PB
* Xplained Nano ATtiny104 and ATtiny416
* Xplained ATxmega256A3BU
* ATxmega128A3U test board (without native USB, but with on-board USB to serial adapter)
* Curiosity Nano ATmega4809, ATtiny3217, AVR128DA48, AVR128DB48, AVR64DD32
* AVR IoT ATmega4808

**Reply by mcuee** [reply]
Wow, that is an impressive list!

**Reply by mcuee** [reply]
Bootloader used for the boards I have.
https://github.com/avrdudes/avrdude/issues/1224#issuecomment-1356499105

**Reply by mcuee** [reply]
@dl8dtl has quite some good tools as well.

For avrdude 7.0 release, he has done the following tests.

* STK500, ISP mode, ATmega8
* STK500, PP mode, ATmega8
* STK500, HVSP mode, ATtiny11 (a device where HVSP is the only option ;-)
* AVRISPmkII, ISP mode, ATmega8
* JTAGICEmkII, ISP mode, ATmega8
* JTAGICEmkII, JTAG mode, ATmega16
* JTAGICEmkII, PDI mode, ATxmega256A3
* USBtinyISP, ATmega8
* Matwei USBISP (STK500v2 clone), ATmega16
* JTAGICE ("mk1") clone, ATmega16
* AtmelICE, JTAG mode, ATmega16
* AtmelICE, ISP mode, ATmega16
* AtmelICE, TPI mode, ATtiny10
* Parallel-port dongle (original "bsd" one from Brian Dean), ATmega8
* "ponyser" serial-port dongle, ATmega8
* FT245R (`ftdi_syncbb`), ATmega8
* FT232R (`ftdi_syncbb`), ATmega8
* UM232H (`avrftdi`), ATmega16
* USBasp, ATmega8
* AVR Dragon, ISP mode, ATmega16
* AVR Dragon, JTAG mode, ATmega16

**Reply by mcuee** [reply]
Got two new tools for avrdude testing (just nice for 8.1 release).
AVR JTAGICE mkII clone (programmer only, no debugging function) -- June 2025
STK600 programmer only clone -- June 2025

**Reply by mcuee** [reply]
In order to better test FreeBSD/OpenBSD/NetBSD, I have set up two old Chuwi mini PCs (each with one extra added 128GB SSD). Sometimes virtual machines are not that good in terms of USB performance. Physical machine is better for testing.

Intel J4125 mini PC, 8GB/256GB +128GB --> triple boot, FreeBSD 14.2, Debian 12, OpenSuse Tumbleweed)
Intel N4200 mini PC, 8GB/256GB +128 GB --> dual boot, NetBSD 10.1 and OpenBSD 7.7.

I have two more Chuwi Intel N100 mini PCs (8GB/256GB configuration with added 128GB SSD) with different Linux distros as well (removing the original installed Windows 11).

**Comment by dl8dtl** [comment]
My main development system is FreeBSD, so for sure, I'm going to test this.
I've also got access to Linux systems including an elderly Raspberry Pi (important for linuxgpio and linuxspi), and Macs both M1 and x86_64.
Enumerating all the tools and devboards I've got access to is way too much here. It would be easier to explain what I don't have :-), e.g. the PICkit4 that was recently mentioned I don't have.
I theory, I could access an old UltraSPARC machine with both FreeBSD and Solaris, but it's quite dated, and can hardly be revived.

**Reply by mcuee** [reply]
@MCUdude 
I think you have quite some programmers and tools as well, in particular, you have the official Microchip PICKit 4 and SNAP as well as AVR Dragon and genuine STK500 kits.

**Reply by mcuee** [reply]
> In theory, I could access an old UltraSPARC machine with both FreeBSD and Solaris, but it's quite dated, and can hardly be revived.

Wow, avrdude even supports Solaris. Solaris does have libusb-1.0/libusb-0.1 support (they also have openusb) for Solaris x86, no idea about Solaris UltraSpac. So not sure anyone have tried the USB programmers there.

**Reply by dl8dtl** [reply]
> Wow, avrdude even supports Solaris.

At least, it used to. I haven't tested the current release there.

> [UltraSparc] So not sure anyone have tried the USB programmers there.

I'm pretty sure I once tested at least an AVRISPmkII on the UltraSPARC platform. After all, these blade server machines don't have any other connectors but USB. (Despite, Solaris even offers a well-defined parallel-port API, so we support parallel-port adapters on Solaris, like we do on Linux and FreeBSD. Probably no longer relevant these days. :-))

**Comment by stefanrueger** [comment]
> It will be good to create a list of testing facilities

Brilliant idea. This is very, very helpful. 

Automation is key. Something along the lines of "tricky" .hex files (with holes, etc) for flash upload/download (one of each relevant class of AVR?), automated EEPROM r/w tests, etc.

**Reply by dl8dtl** [reply]
> Automation is key.

Do you think about anything that integrates into Github's CI?

That's probably going to be pretty hard for the entire amount of hardware (AVR devices vs. programmers) in question. However, I could probably set a Raspberry Pi aside and connect a bit of stuff to it, that can then be attached to the CI. I have no experience with these processes though.

**Comment by stefanrueger** [comment]
> Do you think about anything that integrates into Github's CI?

I haven't thought that far! In the first instance I thought of neat little shell scripts that can test paged flash and eeprom r/w, byte read/write for non flash (or when it's the only method), maybe even write a few safe fuse bytes under carefully choreographed scenarios etc.

But yes, the thought of an array of 45 programmers times 5 different MUCs (one per family) that are always connected to an RPi and will rush into action every time someone commits so much as removing a typo in a comment has something! It's just 225 devices, and with an estimated power of 2 W/device will "only" cost a few thousand USD/GBP/EUR per year to run... :) OTOH, if there was a clever way of multiplexing a dozen of key programmers with 5 MCUs through an interface that the RPi can control then this would be the most wonderful whacky (in a good sense) CI that certainly I've come across...

**Comment by mcuee** [comment]
It is always difficult to carry out CI when HW is involved. There are similar discussions in libusb, libusbdotnet and pyusb.

Ref: pyusb discussion
https://github.com/pyusb/pyusb/issues/235

In the end, emulated USB device is kind of used in libusb.
https://github.com/libusb/libusb/blob/master/tests/umockdev.c

So one way is to develop a "Software emulated hardware" to test avrdude, but it is probably not easy.

**Comment by mcuee** [comment]
I am located in Singapore and I know Chinese so I usually buy stufff from Taobao (part of AliBaba group). 

For example I bought the JTAG ICE clone from Taobao for about US$7.64 (RMB 45 for the unit + 7% GST for Singapore + RMB 3 sea shipment = RMB 51.15 = about US$7.64 ).  Sea shipment is cheaper but it will take me about 1 month to get the stuff. So sometimes I use local online stores like Shopee or Lazada, which has faster shipment (sometimes the stock is already in Singapore, sometimes it is from China).

You can probably buy from AliExpress (part of Alibaba group) which offers global shipment. Or you can look at other places like Amazon.
The AliExpress links below are only for reference as I do not have experiences with the shops. You may need to look at the reviews closely to select the shop to buy.

A quick search shows around US$7 to US$20 for the JTAG ICE clone. My one is the square box type and not the usbasp type.
https://www.aliexpress.com/w/wholesale-jtag-ice-avr.html
https://www.aliexpress.com/item/32222264745.html

Similar search for AliExpress for AVRISP mkii. My one is at about US$15.
https://www.aliexpress.com/w/wholesale-avrisp-mkii.html
https://www.aliexpress.com/item/32960161527.html

The FT232RL with AVR ISP pin header (blue one) is at US$1.44 plus shipment.
https://www.aliexpress.com/item/32273550144.html

**Comment by mcuee** [comment]
For ATxemga break-out or development board, there are not many choices. I found the following from mcuzone. I have bought the ATxmega32D4 mini board from their Taobao shop (which has more options than their AliExpress shop) but I have not figured out how to progam using the PDI interface with the AVR Dragon or the AVRISP mkii clone.
https://wildchip.aliexpress.com/store/group/ATXMEGA-mini-Board/5881453_517582356.html
https://wildchip.aliexpress.com/store/group/ATXMEGA-Dev-Kit/5881453_517927515.html

**Comment by mcuee** [comment]
Nice link here for DIY programmers, including various updi versions like serialupdi, updi-HV, and also usbasp/usbtinyisp
https://github.com/wagiminator/AVR-Programmer

1) usbasp seems to be the most popular low cost USB programmers. There seem to be add-ons to support PDI and maybe updi.
Ref: https://github.com/avrdudes/avrdude/issues/1024

There are many usbasp clones out there -- mostly based on the original design using Atmega8/8A, but there are others using ATmega88/88A. In particular, there are Chinese clones (usually called USBISP) which are using ATmega88 and present as USB HID device. Those clones are using ProgISP and not compatible avrdude. However, some of them can be upgraded using the FW here (USB composite FW with WinUSB and HID). I was able to change one of them to use usbasp FW. But the other two clone failed. You need to short two test PCB vias and then use another programmer to upgrade the FW.
https://github.com/dioannidis/usbasp/tree/master/bin/firmware (usbasp.atmega88.2011-05-28.hex)

2) usbtinyisp is also popular. There are clones using ATtiny2313 (usually discrete DIP part, with 10pin and 6-pin ISP connector) and then there are clones using ATtiny44 and ATtiny45.

3) SerialUPDI/JTAG2UPDI can be made cheaply from USB to TTL converters by adding a few components.
https://github.com/SpenceKonde/AVR-Guidance/blob/master/UPDI/jtag2updi.md

**Comment by mcuee** [comment]
STK500 schematics are openly available so there are also many clones (usually just the programmer portion). Very few of them will support HVSP/HVPP.
Ref: https://mechatronics.me.wisc.edu/labresources/schematicsNManuals.htm

But I can see there are different clones claimed to be STK500v2 compatible but may not exactly the same (some of them do not even use AT90S8535/ATmega8535).

Waveshare seems to be the more famous clone for STK500v2/AVRISP (ISP portion, no HV programming)
https://www.waveshare.com/wiki/USB_AVRISP_CH

Then there are various clone schematics for AVR JTAG ICE.
https://scienceprog.com/build-your-own-avr-jtagice-clone/
https://github.com/vyivanov/AVR-JTAG-ICE-mkI

There are also clones of AVRISP mkii. For example, Waveshare AVRISP XPII seems to be the more famous one. But there are cheap ones which seem to work fine as well.
https://www.waveshare.com/wiki/USB_AVRISP_XPII

There are also clones of JTAG ICE mkii (not many). But the price is not cheap. 
https://www.waveshare.com/wiki/USB_AVR_JTAGICE_XPII

MCUzone has clones for STK500v2, JTAG ICE, and JTAG ICE mkii. They seem to have the cheapest JTAG ICE mkii clone at RMB480 plus GST + shipment (about RMB516.6 or US$77 to reach me by sea shipment).
http://www.mcuzone.com/index.htm
http://www.mcuzone.com/shop/?gallery-18.html (in Chinese)

Probably it is good enough to own an official PICKit 4 at US$77 (or SNAP at US$34.09) than to look at JTAG ICE mkii clone. Most of the people will probably not need Atmel-ICE or Power Debugger.
https://microchipdeveloper.com/snap:versus-pk4

**Comment by mcuee** [comment]
@MCUdude wrote here:
https://github.com/MCUdude/avrdude/pull/3#issuecomment-1183731597

>  Off topic: Do you think Microchip would either donate or let one of us spare time Avrdude developers borrow some hardware to ensure that Avrdude properly supports the hardware? There's a few issues we can't work on without testing it on actual hardware:
> 
> * The Power Debugger hasn't gotten much love, and there's a few features that could be added to Avrdude terminal mode.
> * Potential issue ([#404](https://github.com/avrdudes/avrdude/issues/404)) where xmegas with 256kB flash or more fails if the program is greater than 128kB in size
> * Unconfirmed issue with A4U devices ([#289](https://github.com/avrdudes/avrdude/issues/289))
> * 256A3BU hangs, when programming or reading device signature ([#337](https://github.com/avrdudes/avrdude/issues/337))

**Reply by mcuee** [reply]
https://github.com/avrdudes/avrdude/discussions/1003#discussioncomment-2985205
> If anyone wants to tackle those issues, I guess it's easy enough to convince Microchip to donate a devboard.

@dl8dtl Not so sure if you know someone at Microchip who can help @MCUdude to get the Power Debugger.

**Reply by mcuee** [reply]
@MCUdude 
For the xmega, Microchip seems to only offer `XMEGA A1U Xplained Pro` with ATxmega128A1U and `XMEGA-E5 XPLAINED` with ATxmega32E5. Probably they are not good enough for #337 and #404. 

Not so sure if they have some in-house boards with ATxmega256A3U or similar (for #337 and #404) and ATxmega128A4U or similar (for #289). Probably we have to buy from places like AliExpress.

Edit: Actually Microchip has the ATxmega256A3BU based dev board.
https://www.microchip.com/en-us/development-tool/ATXMEGAA3BU-XPLD

**Reply by mcuee** [reply]
Reference: xmega related unconfirmed issues.
https://github.com/avrdudes/avrdude/discussions/1008#discussioncomment-3117733

**Reply by mcuee** [reply]
Google shows that AliExpress has the ATxmega boards available.
Reference: ATxmega256A3U development board, roughly US$56 + Shipment + tax.
https://www.aliexpress.com/item/4001166867043.html

BTW, the same shop also has the cheaper ATxmega32A4 breakout board.
https://wildchip.aliexpress.com/store/group/ATXMEGA-mini-Board/5881453_517582356.html

**Reply by MCUdude** [reply]
@mcuee I've been in touch with Jan Egil, and it looks like I'll receive some test gear sometime in August/September. I'm not sure what I'll receive though.

**Reply by mcuee** [reply]
That is great.

**Comment by mcuee** [comment]
The following page is a good summary of AVR programmers (in German but Google Translate seem to work well).
https://www.mikrocontroller.net/articles/AVR_In_System_Programmer

**Comment by mcuee** [comment]
I have never tried to use HV programming previously. Today I tried to use it using my STK500 clone (with HVPP sockets for M16/M32 and M8).

```
PS C:\work\avr\avrdude_test\avrdude_bin> .\avrdude -c stk500pp -P COM3 -p m32a -B1 -U flash:r:m32demo_usb_stk500pp_read.hex:i

avrdude.exe: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.01s

avrdude.exe: Device signature = 0x1e9502 (probably m32a)
avrdude.exe: reading flash memory ...

Reading | ################################################## | 100% 4.00s

avrdude.exe: writing output file m32demo_usb_stk500pp_read.hex

avrdude.exe done.  Thank you.

PS C:\work\avr\avrdude_test\avrdude_bin> .\avrdude -c stk500pp -P COM3 -p m32a -B1 -e

avrdude.exe: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.01s

avrdude.exe: Device signature = 0x1e9502 (probably m32a)
avrdude.exe: erasing chip

avrdude.exe done.  Thank you.

PS C:\work\avr\avrdude_test\avrdude_bin> .\avrdude -c stk500pp -P COM3 -p m32a -B1 -U flash:w:m32demo_usb_stk500pp_read.hex:i

avrdude.exe: AVR device initialized and ready to accept instructions

Reading | ################################################## | 100% 0.01s

avrdude.exe: Device signature = 0x1e9502 (probably m32a)
avrdude.exe: NOTE: "flash" memory has been specified, an erase cycle will be performed
             To disable this feature, specify the -D option.
avrdude.exe: erasing chip
avrdude.exe: reading input file m32demo_usb_stk500pp_read.hex for flash
avrdude.exe: writing 292 bytes flash ...

Writing | ################################################## | 100% 0.07s

avrdude.exe: 292 bytes of flash written
avrdude.exe: verifying flash memory against m32demo_usb_stk500pp_read.hex

Reading | ################################################## | 100% 0.05s

avrdude.exe: 292 bytes of flash verified

avrdude.exe done.  Thank you.
```

**Comment by mcuee** [comment]
For DIY UPDI programmers, either jtag2updi or serialupdi can be a good choices.
**jtag2updi (please use latest version)**
https://github.com/ElTangas/jtag2updi

**serialupdi**
https://github.com/SpenceKonde/AVR-Guidance/blob/master/UPDI/jtag2updi.md

One of the better design of serialupdi programmer is from @MCUdude. He has also the **microupdi** programmer.
https://github.com/MCUdude/MegaCoreX#updi-programmers

**Reply by mcuee** [reply]
Basically for average avrdude users, I think usbasp, usbtinyisp, stk500v2 clone or avrispmkii clone will be good one to get started with classic AVRs. avrispmkii also works with xmega. Of course, cheaper ones like FT232RL based programmer can be a good choice as well (a bit slow). Cheap Arduino Uno/Nano clones can be made into good progammers as well (Arduino as ISP, AVR910 or STK500v2).

For UPDI, Microchip SNAP, PICKit 4, jtag2updi (using Uno or Nano clones) and serialupdi are good ones to have. 

Down the road, probably Microchip SNAP and PICKit 4 will be the preferred official tool for all AVRs. usbasp, usbtinyisp, jtag2updi and serialupdi will be the good unofficial tools to have.

**Comment by mcuee** [comment]
I have a AVRISP mkii which did not work. Today I used Microchip Studio to upgrade the FW from 1.23 to 1.24 and now it works.

```
PS C:\work\avr\avrdude_test\avrdude_bin> .\avrdude -c avrisp2 -p m328pb -v

avrdude: Version 7.0-20221221 (292f1d5)
         Copyright the AVRDUDE authors;
         see https://github.com/avrdudes/avrdude/blob/main/AUTHORS

         System wide configuration file is C:\work\avr\avrdude_test\avrdude_bin\avrdude.conf

         Using Port                    : usb
         Using Programmer              : avrisp2
avrdude: usbdev_open(): found AVRISP mkII, serno: FFFFFFFFFFFF
         AVR Part                      : ATmega328PB
         Chip Erase delay              : 10500 us
         PAGEL                         : PD7
         BS2                           : PC2
         RESET disposition             : possible i/o
         RETRY pulse                   : SCK
         Serial program mode           : yes
         Parallel program mode         : yes
         Timeout                       : 200
         StabDelay                     : 100
         CmdexeDelay                   : 25
         SyncLoops                     : 32
         PollIndex                     : 3
         PollValue                     : 0x53
         Memory Detail                 :

                                           Block Poll               Page                       Polled
           Memory Type Alias    Mode Delay Size  Indx Paged  Size   Size #Pages MinW  MaxW   ReadBack
           ----------- -------- ---- ----- ----- ---- ------ ------ ---- ------ ----- ----- ---------
           eeprom                 65    20     4    0 no       1024    4      0  3600  3600 0xff 0xff
           flash                  65     6   128    0 yes     32768  128    256  4500  4500 0xff 0xff
           lfuse                   0     0     0    0 no          1    1      0  4500  4500 0x00 0x00
           hfuse                   0     0     0    0 no          1    1      0  4500  4500 0x00 0x00
           efuse                   0     0     0    0 no          1    1      0  4500  4500 0x00 0x00
           lock                    0     0     0    0 no          1    1      0  4500  4500 0x00 0x00
           signature               0     0     0    0 no          3    1      0     0     0 0x00 0x00
           calibration             0     0     0    0 no          1    1      0     0     0 0x00 0x00

         Programmer Type : STK500V2
         Description     : Atmel AVR ISP mkII
         Programmer Model: AVRISP mkII
         Hardware Version: 1
         Serial number   : FFFFFFFFFFFF
         Firmware Version Controller : 1.23
         Vtarget         : 5.1 V
         SCK period      : 1.00 us

avrdude: AVR device initialized and ready to accept instructions
avrdude: device signature = 0x1fd517
avrdude main() error: expected signature for ATmega328PB is 1E 95 16
        double check chip or use -F to override this check

avrdude done.  Thank you.

PS C:\work\avr\avrdude_test\avrdude_bin> .\avrdude -c avrisp2 -p m328pb -v

avrdude: Version 7.0-20221221 (292f1d5)
         Copyright the AVRDUDE authors;
         see https://github.com/avrdudes/avrdude/blob/main/AUTHORS

         System wide configuration file is C:\work\avr\avrdude_test\avrdude_bin\avrdude.conf

         Using Port                    : usb
         Using Programmer              : avrisp2
avrdude: usbdev_open(): found AVRISP mkII, serno: FFFFFFFFFFFF
         AVR Part                      : ATmega328PB
         Chip Erase delay              : 10500 us
         PAGEL                         : PD7
         BS2                           : PC2
         RESET disposition             : possible i/o
         RETRY pulse                   : SCK
         Serial program mode           : yes
         Parallel program mode         : yes
         Timeout                       : 200
         StabDelay                     : 100
         CmdexeDelay                   : 25
         SyncLoops                     : 32
         PollIndex                     : 3
         PollValue                     : 0x53
         Memory Detail                 :

                                           Block Poll               Page                       Polled
           Memory Type Alias    Mode Delay Size  Indx Paged  Size   Size #Pages MinW  MaxW   ReadBack
           ----------- -------- ---- ----- ----- ---- ------ ------ ---- ------ ----- ----- ---------
           eeprom                 65    20     4    0 no       1024    4      0  3600  3600 0xff 0xff
           flash                  65     6   128    0 yes     32768  128    256  4500  4500 0xff 0xff
           lfuse                   0     0     0    0 no          1    1      0  4500  4500 0x00 0x00
           hfuse                   0     0     0    0 no          1    1      0  4500  4500 0x00 0x00
           efuse                   0     0     0    0 no          1    1      0  4500  4500 0x00 0x00
           lock                    0     0     0    0 no          1    1      0  4500  4500 0x00 0x00
           signature               0     0     0    0 no          3    1      0     0     0 0x00 0x00
           calibration             0     0     0    0 no          1    1      0     0     0 0x00 0x00

         Programmer Type : STK500V2
         Description     : Atmel AVR ISP mkII
         Programmer Model: AVRISP mkII
         Hardware Version: 1
         Serial number   : FFFFFFFFFFFF
         Firmware Version Controller : 1.24
         Vtarget         : 5.1 V
         SCK period      : 8.00 us

avrdude: AVR device initialized and ready to accept instructions
avrdude: device signature = 0x1e9516 (probably m328pb)

avrdude done.  Thank you.
```

**Comment by mcuee** [comment]
My main host side stuff.
Updated on 3-Oct-2023.

1) macOS ARM64: 
Mac Mini M1 2021.  My old Mac Mini 2011 was dead in 2019 so I cannot test old macOS releases.

2) Windows 11 and Linux x86_64: 
Acer Swift 3 laptop, 2021 model -- dual boot, Windows 11 and Ubuntu Linux 20.04 (Intel Core i5-1135G7, 16GB RAM, 512GB SSD)
I have FreeBSD/OpenBSD/NetBSD VMs using VirtualBox as well but sometimes USB stops working inside the VM after upgradiing VirtualBox. 

3) ARM Linux using ARM SBCs
I usually use my Raspberry Pi 400 as it is faster than my other ARM SBCs. I have Raspberry Pi 3B+ and Raspberry Pi 2 (early version) as well and a few other OrangePi boards. This runs Raspberry Pi OS (32bit/64bit) or Ubuntu 64bit Linux.

4) Linux x86_64 physical machine, and Linux/BSD VMs: using mini PC powered by low power Intel CPUs.

Intel J4105 and Intel J4125 Mini PC running Proxmox Virtual Environment (PVE), Linux/BSD VMs and LxC Linux Containers. 
Intel N4200 mini PC running dual boot Ubuntu 22.04 and Debian 12.
I have also ordered two new Intel N100 based Mini PCs.

The Intel J4105 mini PC has 4 gigabit ports and I play with home networking stuff using it with a WiFi 6 Access Point attached to it. You can see that I am running OpenWRT now and two LxC conainers (one with Pi-hole and the other with Adguard Home).

But it also runs BSD and Linux VMs so that I can carry out tests of avrdude and other open source projects as well. USB passthrough works better in PVE than VirtualBox. I am slowly building up PVE8 on the Intel J4125 mini PC (single 2.5Gbe adapter).

![Screenshot 2023-10-07 124411](https://github.com/avrdudes/avrdude/assets/1294740/2556f026-b897-4689-ad08-5feca03e5d54)

**Reply by mcuee** [reply]
Updated on 10-May-2026

1) Main laptops
Acer Swift Go 14 2024 model, bought in 2024, dual boot Windows 11 and Ubuntu 26.04.
MacBook Neo 2026 model, newly bought in 2026, running macOS
Acer Aspire 14 AI 2025 model, newly bought in 2026, running Windows 11 on ARM64.

2) Desktop
AMD Ryzen 5 5600G based desktop, bought in 2024, dual boot Windows 11 and Ubuntu 24.04

3) Mini PCs -- installed FreeBSD/OpenBSD/NetBSD physical machine so that I can test them on physical machines.

4) Raspberry Pi 500 -- bought in early 2025

**Comment by mcuee** [comment]
Windows 11 on ARM64 is not recommended for avrdude users -- many drivers may not be compatible. 
1) Microchip also does not provide drivers for Windows on ARM64.
2) Some USB to Serial/TTL adapters may not work due to lack of driver.

Good thing:
1) Generic USB device can use the following trick to install WinUSB driver even though Zadig does not work. So libusb and libusb-compat-0.1 may still work.
* https://github.com/pbatard/libwdi/issues/289#issuecomment-3680073634

2) Generic USB HID device should still work. So HIDAPI should still work.

**Comment by mcuee** [comment]
macOS for Apple Silicon now has native container support which may help to check Linux building without using a VM.
https://github.com/apple/container

No support of USB Passthrough yet.
* https://github.com/apple/container/issues/640

```
mcuee in ~ % container run --rm -it debian:13 bash
root@b502d5fa-00fe-4dfa-8cbb-86d853ff7278:/# cd
...
...
root@b502d5fa-00fe-4dfa-8cbb-86d853ff7278:~/avrdude# avrdude --version
Avrdude version 8.2-20260723 (a2f944f6)
Copyright see https://github.com/avrdudes/avrdude/blob/main/AUTHORS
Use https://github.com/avrdudes/avrdude/issues to report bugs and ask questions
root@b502d5fa-00fe-4dfa-8cbb-86d853ff7278:~/avrdude# exit
exit
mcuee in ~ %
```

**Reply by mcuee** [reply]
Microsoft has WSL Container as well and it can use usbipd-win to access USB devices.
https://learn.microsoft.com/en-us/windows/wsl/wsl-container
https://learn.microsoft.com/en-us/windows/wsl/connect-usb
